spec/lib/ruby_require_leak_spec.rb

```diff
@@ -0,0 +1,45 @@
+describe "Ruby 'require' leak with Pathnames" do
+  context "to prevent" do
+    # If this test fails, then either this repo, or one of the provider repos
+    # has added a entry to `config.autoload_paths` that is a Pathname instead
+    # of a raw string.  Until we upgrade to ruby >= 2.5, all Pathnames in the
+    # autoload_paths/$LOAD_PATH should be converted to Strings.
+    #
+    # You can use the code in the expect block as a way of testing this in a
+    # `bin/rails console`:
+    #
+    #     irb> $LOAD_PATH.select { |p| p.class == Pathname }
+    #     #=> []
+    #
+    # More info can be found here:
+    #
+    #     * https://bugs.ruby-lang.org/issues/14372
+    #
+    it "has no Pathnames in the $LOAD_PATH" do
+      expect($LOAD_PATH.select { |p| p.class == Pathname }).to be_empty
+    end
+  end
+
+  context "this file can be removed when" do
+    # If this test fails, then the ManageIQ project has been upgraded to a
+    # version of ruby that no longer requires converting Pathnames in the
+    # autoload_paths/$LOAD_PATH to strings.  This means this file can be
+    # deleted and the changes in the the config/application.rb and
+    # lib/manageiq/**/engine.rb files can be reverted. See the following PRs
+    # for examples of what needs to be reverted:
+    #
+    #     * https://github.com/ManageIQ/manageiq/pull/16837
+    #     * https://github.com/ManageIQ/manageiq-api/pull/288
+    #     * https://github.com/ManageIQ/manageiq-automation_engine/pull/146
+    #     * https://github.com/ManageIQ/manageiq-ui-classic/pull/3266
+    #     * https://github.com/ManageIQ/manageiq-graphql/pull/34
+    #
+    it "is running on ruby >= 2.5.0" do
+      current_ruby_version = Gem::Version.new(RbConfig::CONFIG["ruby_version"])
+      ruby_2_5_0_version   = Gem::Version.new("2.5.0")
+      err_msg              = "expected ruby to be less than 2.5.0"
+
+      expect(current_ruby_version).to be < ruby_2_5_0_version, err_msg
```

Comment:
I'm not sure I like this test because it will fail if we run Ruby 2.5 locally.  I would think a simple STDERR.puts warning in an initializer or at the top of Gemfile would be sufficient.

Reply:
Yeah, but that just annoying too, and isn't going to stop a build, so it might just become ignored noise after a while.

That said, I see your point about it running locally.  Maybe wrap this with a `if ENV["CI"]` ?

Reply:
Also, I am not a big fan of shipping stuff like this into something that is run in prod.  I am cool with it being in a test, since those shouldn't be shipped technically, though, I think we still include them anyway.

Reply:
Since it's for this file, we can puts at the top of this file.  The objective is to remind us about this hack we did.  Maybe by the time that all of us are running ruby 2.5.0 fairly often, we'll know which way we want to go: Strings or Pathnames.  I'm leaning towards Strings right now because of what I've seen in testing Pathnames in $LOAD_PATH.  I've always had an understanding that require should be nearly a no-op if the file is already required but it's clearly not with Pathnames in the path.

Reply:
I wouldn't mind if ENV["CI"]

Reply:
Though to @kbrock's point, when 2.4 becomes the minimum, we will run the specs against 2.4 and 2.5, and this will always fail for 2.4.

Reply:
> Though to @kbrock's point, when 2.4 becomes the minimum, we will run the specs against 2.4 and 2.5, and this will always fail for 2.4.

@Fryguy I assume you mean "will always fail for 2.5" here, correct?

Good point, would wrapping it with `if ENV["CI"] and ENV["TRAVIS_ALLOW_FAILURE"] == "true"` handle that better than?  See docs:  https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables

Reply:
> ...we'll know which way we want to go: Strings or Pathnames. I'm leaning towards Strings right now...

@jrafanie yeah, I am thinking I am going to change the copy a bit on these comments while I am at it, to say more like "we can re-evaluate once we are on 2.5".

Reply:
I think the concerns from here should be addressed now.  Added a `if ENV["CI"] and ENV["TRAVIS_ALLOW_FAILURE"] != "true"` block so this will only fail when we have full converted to something greater than or equal to ruby 2.5.0 on CI.  Testing locally should work just fine.